### DISCUSSION TITLE
Customisation of Code Completion

### DISCUSSION THREAD

**Discussion by markusvoelter** [eclipse-langium/langium#712] [Q&A]
I am working on customizing code completion for my language. In one particular parser rule (AST type) I need to code complete a string into an external configuration (think: XML file). I got the fundamentals to work by writing my own completion provider and extending the default one. I overwrote the completeForRule method with my own logic and it basically works. One question
however:

Let's assume the completions (based on my external XML file) are ABC, AXX and DEF. If I press Ctrl-Space when the user has already written A, then I'd assume to only get ABC and AXX as proposals. Is this behavior intended to be default (ie, handled outside my completeForRule method) or should I implement this myself? If the latter, it seems that the completeForRule method doesn't get enough arguments; prefix or something is missing.

What do I do? @msujew, is your improvement of the code completion provider addressing this (@montymxb hinted at that)?

**Comment by spoenemann** [comment]
Hi Markus! AFAICT the default completion provider should take care of filtering proposals to match the current prefix:
https://github.com/langium/langium/blob/7305814026c17348df65e5d13cb9daaba8fd2437/packages/langium/src/lsp/completion/completion-provider.ts#L109

Can you confirm this from the editor behavior?

**Reply by markusvoelter** [reply]
Hi Miro.

No, I cannot confirm. Here is the list without prefix:

![Screenshot 2022-10-05 at 12 44 01](https://user-images.githubusercontent.com/592330/194042570-14f273b9-b176-400b-93af-e773161fecfd.png)

... and here it is when I have typed "I", the start of two of the proposals:

![Screenshot 2022-10-05 at 12 44 12](https://user-images.githubusercontent.com/592330/194042619-54f39dd3-14a1-4759-a96a-b80e1dab125a.png)

**Reply by markusvoelter** [reply]
In fact, the text "items.filter" are not in the DefaultCompletionProvider I am extending. I use langium "0.5.0-next.f2b3802"

**Reply by spoenemann** [reply]
We released v0.5.0 today, could you try with that version?

**Reply by markusvoelter** [reply]
Ah, ok :-) Will do.

**Reply by markusvoelter** [reply]
After updating, the completion provide I wrote doesn't provide any completions anymore. Will have to investigate what changed.

**Reply by markusvoelter** [reply]
@spoenemann the completionFor() method that I have overridden is never called:

![image](https://user-images.githubusercontent.com/592330/194067288-f8368101-3b84-4357-b046-1b8db30c6b5c.png)

What am I misunderstanding?

**Reply by markusvoelter** [reply]
Not even the getCompletion method is called if I overwrite it. My completion provider is "loaded" though.

![image](https://user-images.githubusercontent.com/592330/194075082-33b105c1-2cc8-4462-b9a1-899f9f1c060e.png)

**Reply by markusvoelter** [reply]
One more data point: I have set up a complete new hello world project based on your Getting Started guide. Default code completion works with my grammar. I then implemented my own dummy completion provider:

![image](https://user-images.githubusercontent.com/592330/194123263-5dfe595e-a785-4095-b26d-441c3ebcc666.png)

and registered it:

![image](https://user-images.githubusercontent.com/592330/194123343-8e7c4261-210b-451b-b5cb-8aeafbdaa077.png)

Doesn't work - my custom code completion proposals are not shown.

**Reply by msujew** [reply]
Oh, we forgot to mention on the changelog that the `lsp.completion` module does no longer exist. It's only `lsp.CompletionProvider` now:

https://github.com/langium/langium/blob/7305814026c17348df65e5d13cb9daaba8fd2437/packages/langium/src/default-module.ts#L64-L65

**Reply by markusvoelter** [reply]
Ok thanks, @msujew, code completion is working again. However, two problems:

First, he prefix-filtering does not work. I get an empty completion menu as soon as I complete with a prefix. It seems to me that the matching logic no longer works; when I start typing, then the parser already recognises the next rule / ast node, and my logic in completeFor no longer works. I understand why this is the case, but it seems to me that the post-logic prefix filtering just does not work in principle: it suggests, "do what you did before, and then filter".

Second, when I create a proposal like this 

![image](https://user-images.githubusercontent.com/592330/194259486-19f4ef31-6ae3-4b11-8f24-20ba8d24c958.png)

then I would assume that after selecting k in the code completion menu, the editor inserts k {}, as given in the insertText. However, this is consistently not happening, it inserts only k.

**Reply by markusvoelter** [reply]
@msujew Here is an example of the first problem. 

![image](https://user-images.githubusercontent.com/592330/194271677-99805e76-eac3-4dac-8a73-a9c2a07f4894.png)

If I start with the "n" here, then the API recognises astNode = Child (a name of an AST Type) and next = "Child". What (I think) it should provide to the API is astNode = Node (another AST type in my language), next = "Child" (because I have just started typing the prefix of what will then *become* a Child object. The parser of course starts recognising the "n" as the (full) name of a child, and then claims we are in one. As far as I can tell, the arguments of completeFor do not allow me to recognise this situation and react accordingly.

**Reply by msujew** [reply]
@markusvoelter For the second problem: `insertText` is generally not recommended to be used and will actually be ignored by vscode, since we instead provide a `textEdit` for the completion item. See [here](https://github.com/langium/langium/blob/fa1dbba52a698a2c60191bcb4caffa6fd0b9b347/packages/langium/src/lsp/completion/completion-provider.ts#L264-L268). It is generally recommended to use `textEdit` instead of `insertText`, since it also allows us to provide positional information in addition to the text that we want to add to the document. Though it might make sense to change the behavior to automatically pass in the `insertText` into the `textEdit` if one is provided.

For the first problem: We perform a two-pass completion where we first identify which elements could appear [before the word at the current position](https://github.com/langium/langium/blob/fa1dbba52a698a2c60191bcb4caffa6fd0b9b347/packages/langium/src/lsp/completion/completion-provider.ts#L82-L87). Every completion result from this first pass is then filtered by the preceding text and the [second pass](https://github.com/langium/langium/blob/fa1dbba52a698a2c60191bcb4caffa6fd0b9b347/packages/langium/src/lsp/completion/completion-provider.ts#L104-L116) then takes all tokens up to the cursor position into account. This seems to work correctly in our projects, see the [langium-lox](https://github.com/langium/langium-lox) language for example:

![image](https://user-images.githubusercontent.com/4377073/194724448-6997cc06-0f74-47ef-97e8-bde968d2c868.png)

Do you mind sharing a reproducible example where the completion doesn't behave as expected/outlined?

**Reply by markusvoelter** [reply]
Thanks @msujew. Regarding the second problem: the editText works. A suggestion for API improvement: if I overwrite completionFor (or completionForRule) then the CompletionParams are not available. I stored them in a member of my Completion Provider, but it would be nicer (and more obvious how to use) if the DefaultCompletionProvider would pass the params down to these methods.

**Reply by markusvoelter** [reply]
More problems/questions, @msujew, unfortunately. If I just pass the completion menu entry to the acceptor (no partial item), then the prefix thing works:

![image](https://user-images.githubusercontent.com/592330/194820418-a1d5f436-e510-4c8b-a4e4-072bbe0f021d.png)
![image](https://user-images.githubusercontent.com/592330/194820457-90a3cca5-4d1a-46e1-9fa7-18f0a25f4374.png)

But if I pass an editText, then the prefix is not subtracted from the inserted text. Looks like here you should also filter for the editText:

![image](https://user-images.githubusercontent.com/592330/194820917-413bcd3a-0c24-4cb4-9ac5-b2a24517974a.png)

If this shouldn't happen automatically, then I need access to that prefix (missingPart, I guess) at the location where I calculate the editText entries (again, in completionFor and completionForRule).

**Reply by markusvoelter** [reply]
I duplicated the DefaultCompletionProvider and adapted that filter call to also check for the e.editText.newText. Now the filtering works, but we're back to ignoring the newText upon insert (only the item label is inserted). Guess I'm missing something here ....

**Reply by markusvoelter** [reply]
Actually, this also happens if I do not adapt the filtering logic: as soon as I use code completion with a prefix, the label is used instead of the textEdit.newText.

**Reply by markusvoelter** [reply]
BTW, I can provide my project, we can also do a screenshare to explore the problem together. Might be the fastest route.

**Reply by markusvoelter** [reply]
ping, @msujew

**Reply by msujew** [reply]
@markusvoelter I've sent you a mail (@acm.org) with details for a screensharing session 👍

**Comment by goto40** [comment]
@markusvoelter Is it possible to share some snippets of your code? I am also playing with the CompletionProvider...
